### PR TITLE
8330684: ClassFile API runs into StackOverflowError while parsing certain class' bytes

### PR DESCRIPTION
**Pull request by asotona**
ClassFile API dives into the nested constant pool entries without type restrictions, while parsing a class file. Validation of the entry is performed post-parsing. Specifically corrupted constant pool entry may cause infinite loop during parsing and throws SOE.
This patch resolves the issue by providing specific implementations for the nested CP entries parsing, instead of sharing the common (post-checking) code.
Added test simulates the situation on inner-looped method reference entry.

Please review.

Thank you,
Adam

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8330684](https://bugs.openjdk.org/browse/JDK-8330684): ClassFile API runs into StackOverflowError while parsing certain class' bytes (**Bug** - P4)

### Reviewers
 * [Paul Sandoz](https://openjdk.org/census#psandoz) (@PaulSandoz - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18907/head:pull/18907` \
`$ git checkout pull/18907`

Update a local copy of the PR: \
`$ git checkout pull/18907` \
`$ git pull https://git.openjdk.org/jdk.git pull/18907/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18907`

View PR using the GUI difftool: \
`$ git pr show -t 18907`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18907.diff">https://git.openjdk.org/jdk/pull/18907.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18907#issuecomment-2071649181)

### CODE REVIEW

**Review by PaulSandoz — COMMENTED** [review]
Rather than duplicating some checks I wonder if it is possible to add a private method `entryByIndex(int index, int expectedTag)` that the existing `entryByIndex` defers to. If the `expectedTag` is non-negative then it checks `tag` against `expectedTag` before proceeding to the switch expression. Then the implementations of `readClassEntry` etc can be adjusted to pass along the expected tag.

**Comment by PaulSandoz** [comment]
> Unfortunately it would have to be an expected tags list or an extra constructed bit mask, due to the multiple tags allowed for MemberRefEntry and it would slightly affect the performance. 

Ah yes, i missed that. It could be two tags, a lower and upper bound, because TAG_FIELDREF, TAG_METHODREF, and TAG_INTERFACEMETHODREF are consecutive values (9 to 11).

**Comment by asotona** [comment]
> It could be two tags, a lower and upper bound, because TAG_FIELDREF, TAG_METHODREF, and TAG_INTERFACEMETHODREF are consecutive values (9 to 11).

OK, I've implemented it with lower and upper bound tags. Thanks!

File: src/java.base/share/classes/jdk/internal/classfile/impl/ClassReaderImpl.java [hunk 1]
```diff
@@ -423,7 +431,11 @@ public PoolEntry readEntry(int pos) {
 
     @Override
     public <T extends PoolEntry> T readEntry(int pos, Class<T> cls) {
-        var e = readEntry(pos);
+        return readEntry(pos, cls, 0, 0xff);
+    }
+
+    private <T extends PoolEntry> T readEntry(int pos, Class<T> cls, int lowerBoundTag, int upperBoundTag) {
```
Comment (ExE-Boss) [inline src/java.base/share/classes/jdk/internal/classfile/impl/ClassReaderImpl.java:437]:
```suggestion
    private <T extends PoolEntry> T readEntry(int pos, Class<T> cls, int expectedTag) {
        return readEntry(pos, cls, expectedTag, expectedTag);
    }

    private <T extends PoolEntry> T readEntry(int pos, Class<T> cls, int lowerBoundTag, int upperBoundTag) {
```

File: src/java.base/share/classes/jdk/internal/classfile/impl/ClassReaderImpl.java [hunk 2]
```diff
@@ -454,27 +466,27 @@ public Utf8Entry readUtf8EntryOrNull(int pos) {
 
     @Override
     public ModuleEntry readModuleEntry(int pos) {
-        return readEntry(pos, ModuleEntry.class);
+        return readEntry(pos, ModuleEntry.class, TAG_MODULE, TAG_MODULE);
```
Comment (ExE-Boss) [inline src/java.base/share/classes/jdk/internal/classfile/impl/ClassReaderImpl.java:469]:
```suggestion
        return readEntry(pos, ModuleEntry.class, TAG_MODULE);
```

File: src/java.base/share/classes/jdk/internal/classfile/impl/ClassReaderImpl.java [hunk 3]
```diff
@@ -454,27 +466,27 @@ public Utf8Entry readUtf8EntryOrNull(int pos) {
...
     }
 
     @Override
     public PackageEntry readPackageEntry(int pos) {
-        return readEntry(pos, PackageEntry.class);
+        return readEntry(pos, PackageEntry.class, TAG_PACKAGE, TAG_PACKAGE);
```
Comment (ExE-Boss) [inline src/java.base/share/classes/jdk/internal/classfile/impl/ClassReaderImpl.java:474]:
```suggestion
        return readEntry(pos, PackageEntry.class, TAG_PACKAGE);
```

File: src/java.base/share/classes/jdk/internal/classfile/impl/ClassReaderImpl.java [hunk 4]
```diff
@@ -454,27 +466,27 @@ public Utf8Entry readUtf8EntryOrNull(int pos) {
...
     }
 
     @Override
     public ClassEntry readClassEntry(int pos) {
-        return readEntry(pos, ClassEntry.class);
+        return readEntry(pos, ClassEntry.class, TAG_CLASS, TAG_CLASS);
```
Comment (ExE-Boss) [inline src/java.base/share/classes/jdk/internal/classfile/impl/ClassReaderImpl.java:479]:
```suggestion
        return readEntry(pos, ClassEntry.class, TAG_CLASS);
```

File: src/java.base/share/classes/jdk/internal/classfile/impl/ClassReaderImpl.java [hunk 5]
```diff
@@ -454,27 +466,27 @@ public Utf8Entry readUtf8EntryOrNull(int pos) {
...
     }
 
     @Override
     public NameAndTypeEntry readNameAndTypeEntry(int pos) {
-        return readEntry(pos, NameAndTypeEntry.class);
+        return readEntry(pos, NameAndTypeEntry.class, TAG_NAMEANDTYPE, TAG_NAMEANDTYPE);
```
Comment (ExE-Boss) [inline src/java.base/share/classes/jdk/internal/classfile/impl/ClassReaderImpl.java:484]:
```suggestion
        return readEntry(pos, NameAndTypeEntry.class, TAG_NAMEANDTYPE);
```

File: src/java.base/share/classes/jdk/internal/classfile/impl/ClassReaderImpl.java [hunk 6]
```diff
@@ -454,27 +466,27 @@ public Utf8Entry readUtf8EntryOrNull(int pos) {
...
     }
 
     @Override
     public MethodHandleEntry readMethodHandleEntry(int pos) {
-        return readEntry(pos, MethodHandleEntry.class);
+        return readEntry(pos, MethodHandleEntry.class, TAG_METHODHANDLE, TAG_METHODHANDLE);
```
Comment (ExE-Boss) [inline src/java.base/share/classes/jdk/internal/classfile/impl/ClassReaderImpl.java:489]:
```suggestion
        return readEntry(pos, MethodHandleEntry.class, TAG_METHODHANDLE);
```

**Review by ExE-Boss — COMMENTED** [review]
Maybe also add a `readEntry` overload which takes a single expected `TAG_*` value for use by `readModuleEntry`, `readPackageEntry`, `readClassEntry`, `readNameAndTypeEntry`, and `readMethodHandleEntry`:

**Review by PaulSandoz — APPROVED** [review]
Very nice. I think we got lucky this worked out :-)

**Pull request closed without merging**